ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -674,38 +684,67 @@ private object OracleQueries extends Queries {
       gvs: Get[Vector[String]],
       pvs: Put[Vector[String]],
   ): T[Query0[DBContract[Mark, JsValue, JsValue, Vector[String]]]] = {
-    val Seq((tpid, predicate @ _)) = queries // TODO SC handle more than one
     println("selecting")
-    import Queries.CompatImplicits.catsReducibleFromFoldable1
-    // % is explicitly reserved by specification as a delimiter
-    val q = sql"""SELECT c.contract_id, key, payload, agreement_text, sd.parties, od.parties
-                  FROM (contract c
-                        LEFT JOIN signatories sm ON (c.contract_id = sm.contract_id)
-                        LEFT JOIN observers om ON (c.contract_id = om.contract_id))
-                       LEFT JOIN (SELECT contract_id, LISTAGG(party, '%') parties
-                                  FROM signatories GROUP BY contract_id) sd
-                              ON (c.contract_id = sd.contract_id)
-                       LEFT JOIN (SELECT contract_id, LISTAGG(party, '%') parties
-                                  FROM observers GROUP BY contract_id) od
-                              ON (c.contract_id = od.contract_id)
-                  WHERE (""" ++ Fragments.in(fr"sm.party", parties) ++
-      fr" OR " ++ Fragments.in(fr"om.party", parties) ++ sql""")
-                        AND tpid = $tpid AND (""" ++ predicate ++ sql")"
+    // we effectively shadow Mark because Scala 2.12 doesn't quite get
+    // that it should use the GADT type equality otherwise
```

Comment:
Could you expand on that? I don’t understand what the issue is and how this fixes it.

Reply:
As a rule Scala is perfectly aware of type equalities in GADTs, i.e. _for all_ arbitrary `F: * -> *`, `F[A] ~ F[B]`. But sometimes it won't _use_ that fact until you give it a nudge. In this code, the "nudge" is the ascription `T[Query0[DBContract[Mark, JsValue, JsValue, Vector[String]]]]`, but this doesn't apply until the _result_ of `queryByCondition`, so letting Scala operate under the fantasy that Int (or SurrogateTpId) is not Mark, until it really must be, is most convenient.

I think they [finally fixed this in Scala 3](https://twitter.com/not_xuwei_k/status/1332751181119184896).